cpp/src/arrow/util/iterator.h

```diff
@@ -28,84 +29,141 @@
 
-      if (value == NULLPTR) break;
+      if (value == end) break;
 
       ARROW_RETURN_NOT_OK(visitor(std::move(value)));
     }
 
     return status;
   }
+
+  bool operator==(const Iterator& other) const { return ptr_ == other.ptr_; }
+
+  explicit operator bool() const { return ptr_ == NULLPTR; }
+
+ private:
+  template <typename HasNext>
+  static void Delete(void* ptr) {
+    delete static_cast<HasNext*>(ptr);
+  }
+
+  static void NoopDelete(void*) {}
+
+  template <typename HasNext>
+  static Status Next(void* ptr, T* out) {
+    return static_cast<HasNext*>(ptr)->Next(out);
+  }
+
+  std::unique_ptr<void, void (*)(void*)> ptr_;
+  Status (*next_)(void*, T*) = NULLPTR;
+
+  friend Iterator IteratorEnd(internal::type_constant<Iterator>) {
+    // end condition for an Iterator of Iterators is a default constructed (null) iterator
+    return Iterator();
+  }
 };
 
-template <typename Fn, typename T>
-class FunctionIterator : public Iterator<T> {
+template <typename Ptr, typename T>
+class PointerIterator {
  public:
-  using IteratorType = Iterator<T>;
-
-  explicit FunctionIterator(Fn fn) : fn_(std::move(fn)) {}
+  explicit PointerIterator(Ptr ptr) : ptr_(std::move(ptr)) {}
 
-  Status Next(T* out) override { return fn_(out); }
+  Status Next(T* out) { return ptr_->Next(out); }
 
  private:
-  Fn fn_;
+  Ptr ptr_;
 };
 
-template <typename Fn, typename T = typename std::remove_pointer<
-                           internal::call_traits::argument_type<0, Fn>>::type>
-std::unique_ptr<FunctionIterator<Fn, T>> MakeFunctionIterator(Fn fn) {
-  return std::unique_ptr<FunctionIterator<Fn, T>>(
-      new FunctionIterator<Fn, T>(std::move(fn)));
+/// \brief Construct an Iterator which dereferences a (possibly smart) pointer
+/// to invoke its Next function
+template <typename Ptr,
+          typename Pointed = typename std::decay<decltype(*std::declval<Ptr>())>::type,
+          typename Fn = decltype(std::mem_fun(&Pointed::Next)),
+          typename T = typename std::remove_pointer<
+              internal::call_traits::argument_type<1, Fn>>::type>
+Iterator<T> MakePointerIterator(Ptr ptr) {
+  return Iterator<T>(PointerIterator<Ptr, T>(std::move(ptr)));
 }
 
-template <typename T>
-class EmptyIterator : public Iterator<T> {
+template <typename Fn, typename T>
+class FunctionIterator {
  public:
-  explicit EmptyIterator(Status s = Status::OK()) : status_(s) {}
+  explicit FunctionIterator(Fn fn) : fn_(std::move(fn)) {}
 
-  Status Next(T* out) override {
-    *out = NULLPTR;
-    return status_;
-  }
+  Status Next(T* out) { return fn_(out); }
 
  private:
-  Status status_;
+  Fn fn_;
 };
 
+/// \brief Construct an Iterator which invokes a callable on Next()
+template <typename Fn, typename T = typename std::remove_pointer<
+                           internal::call_traits::argument_type<0, Fn>>::type>
+Iterator<T> MakeFunctionIterator(Fn fn) {
+  return Iterator<T>(FunctionIterator<Fn, T>(std::move(fn)));
```

Comment:
It doesn't matter here: the two are equivalent *in the case of taking ownership*, since a new instance of Fn is constructed either way.

using forwarding:
```
template <typename T>
void append(vector<T>* vec, T&& t) {
  vec->push_back(std::forward<T>(t));
}

int main() {
  std::string str = "hello";
  append(&vec_of_string, str); // lvalue reference to str passed as argument
  append(&vec_of_string, std::move(str)); // rvalue reference to str passed as argument
}
```


using move:
```
template <typename T>
void append(vector<T>* vec, T t) {
  vec->push_back(std::move(t));
}

int main() {
  std::string str = "hello";
  append(&vec_of_string, str); // parameter is copy constructed from str
  append(&vec_of_string, std::move(str)); // parameter is move constructed from str
}
```